src/hook.rs

```diff
@@ -0,0 +1,133 @@
+use std::{slice, ptr, mem, str};
+use std::borrow::Cow;
+use libc::{self, c_int};
+use ffi::{self, lua_State, lua_Debug};
+use lua::extra_data;
+use util::callback_error;
+
+/// Contains information about the running code at the moments specified when setting the hook.
+/// Each field is described in the [Lua 5.3 documentaton][lua_doc].
+///
+/// Depending on when the hook is called, some fields might not be available. In these cases,
+/// integers and booleans might not be valid and/or strings are set to `None`.
+///
+/// [lua_doc]: https://www.lua.org/manual/5.3/manual.html#lua_Debug
+#[derive(Clone, Debug)]
+pub struct Debug<'a> {
+    pub name: Option<Cow<'a, str>>,
+    pub namewhat: Option<Cow<'a, str>>,
+    pub what: Option<Cow<'a, str>>,
+    pub source: Option<Cow<'a, str>>,
+    pub curr_line: u32,
+    pub line_defined: u32,
+    pub last_line_defined: u32,
+    pub num_ups: u32,
+    pub num_params: u32,
+    pub is_vararg: bool,
+    pub is_tailcall: bool,
+    pub short_src: Option<Cow<'a, str>>
+}
+
+impl<'a> Debug<'a> {
+    /// Construct a new `Debug` structure that is not associated with a Lua debug structure. It
+    /// involves some string copying.
+    pub fn to_owned(&'a self) -> Debug<'static> {
+        Debug {
+            name: self.name.as_ref().and_then(|s| Some(Cow::Owned(s.as_ref().to_string()))),
+            namewhat: self.namewhat.as_ref().and_then(|s| Some(Cow::Owned(s.as_ref().to_string()))),
+            what: self.what.as_ref().and_then(|s| Some(Cow::Owned(s.as_ref().to_string()))),
+            source: self.source.as_ref().and_then(|s| Some(Cow::Owned(s.as_ref().to_string()))),
+            curr_line: self.curr_line,
+            line_defined: self.line_defined,
+            last_line_defined: self.last_line_defined,
+            num_ups: self.num_ups,
+            num_params: self.num_params,
+            is_vararg: self.is_vararg,
+            is_tailcall: self.is_tailcall,
+            short_src: self.short_src.as_ref().and_then(|s| Some(Cow::Owned(s.as_ref().to_string())))
+        }
+    }
+}
+
+/// Indicate in which circumstances the hook should be called by Lua.
+pub struct HookOptions {
+    /// Before a function call.
+    pub calls: bool,
+    /// When Lua returns from a function.
+    pub returns: bool,
+    /// Before executing a new line, or returning from a function call.
+    pub lines: bool,
+    /// After a certain amount of instructions specified by `count`.
+    pub after_counts: bool,
+    /// Specify how many instructions to execute before calling the hook. Only effective when
+    /// `after_counts` is set to true.
+    pub count: u32
```

Comment:
It might make sense to turn `after_counts` into an `Option<u32>` instead of ignoring this field when `after_counts` is `false`